openapi/paths/describe-inputs.yaml

```diff
@@ -0,0 +1,110 @@
+  get:
+    tags:
+    - Workflow Inputs
+    summary: Get workflow Inputs Structure
+    description: This endpoint describes the structure of workflow's inputs as a json schema. Consumers of this endpoint can use the generated schema to properly format their `workflow_params` when submitting a new run.
+    operationId: DescribeWorkflowInputs
+    parameters:
+    - name: workflow_url
+      in: query
+      description: The absolute URL to a workflow file.
+      style: form
+      explode: true
+      schema:
+        type: string
+        format: URL
+    responses:
+      200:
+        description: 'Successfully parsed the workflow and returned a JSON schema for the workflow inputs'
+        headers: {}
+        content:
+          application/json:
+            schema:
+              $ref: 'http://json-schema.org/draft-07/schema#'
+      400:
+        description: The request is malformed.
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/ErrorResponse'
+      401:
+        description: The request is unauthorized.
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/ErrorResponse'
+      403:
+        description: The requester is not authorized to perform this action.
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/ErrorResponse'
+      500:
+        description: An unexpected error occurred.
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/ErrorResponse'
+    deprecated: false
+  post:
+    tags:
+    - Workflow Inputs
+    summary: Get workflow Inputs Structure With Attachments
+    description: >-
+      This endpoint describes the structure of workflow's inputs as a json schema. Consumers of this endpoint can use the generated schema to properly format their `workflow_params` when submitting a new run.
```

Comment:
I actually really like this idea! Another complaint I have about WES is that it provides a mechanism to get default engine params, but not list ALL available engine params, I even went as far as making a suggestion here https://github.com/ga4gh/workflow-execution-service-schemas/issues/158#issuecomment-741075913.

If we were to add support for this I would probably change a few things in the pr:

- instead of `/describe-inputs` change it to `/descripe-params` or just `/parameters`
- Calling this endpoint wihtout a workflow would only show the available workflow parameters
- Calling this endpoint WITH a workflow would show BOTH the available engine params and the schema expected for the inputs
- We could also potentially extend this to optionally include a schema for the outputs

An example response would be:
```http
# No workflow
GET /describe-params
{
  "available_engine_params": {
    "some_field": {
      "description": "This is a field. It can do things",
      "type": "String",
      "default": null,
      "examples": [
         "foo",
         "bar"
       ]
    } 
  }
}

# With workflow
GET /describe-params?workflow_url=foo.wdl
{
  "workflow_parameters": {
     "$id": "workflow/wdl/foo.wdl",
      "$schema": "https://json-schema.org/draft/2020-12/schema",
      "title": "foo.wdl",
      "type": "object",
      "properties": { 
       }
  },
  "available_engine_params": {
    "some_field": {
      "description": "This is a field. It can do things",
      "type": "String",
      "default": null,
      "examples": [
         "foo",
         "bar"
       ]
    } 
  }
}
```

Reply:
What about a WES that supports multiple workflow engines or engine versions, where available parameters may differ?

Reply:
Not sure, whether you intend to also want to be more specific about the response (-- it's not yet in the PR).

**If** you want to be more specific, maybe it would make sense to add a way to add annotations to the engine params (and also workflow params). For workflow params one may want to add EDAM ontology terms for each parameter. Similarly, for engine parameters one may want to use an ontology. Also the multiple-engine/engine-version use case may one way be covered with this (although if all version's and engine's options are returned, the response might get a bit long).

I'm not sure whether simple tags or key-value pairs would be necessary for this. 

Or do you think this should be covered by the "type" field?